### ISSUE TITLE
make_functional doesn't work with parameter sharing

### ISSUE THREAD

**Issue by zou3519** [pytorch/functorch#446]
Something like the following:
```
import torch
import torch.nn as nn
from functorch import make_functional_with_buffers

class Foo(nn.Module):
    def __init__(self):
        super().__init__()
        self.bias = nn.Parameter(torch.randn(3))
        self.linear = nn.Linear(3, 3)
        self.linear.bias = self.bias
        self.linear_cloned = self.linear
        self.register_buffer('buffer', torch.randn(3))
        self.register_buffer('buffer_cloned', self.buffer)

mod = Foo()

func, params, buffers = make_functional_with_buffers(mod)
```

This is relatively common and should be fixable with @Chillee's https://github.com/pytorch/pytorch/pull/71542

**Comment by zou3519** [comment]
NB: this might be blocked on the resolution of https://github.com/pytorch/pytorch/pull/71542

**Comment by zou3519** [comment]
Initial resolution was to raise a nice error message and we wait for people to tell us they ran into it :)

**Comment by xhan77** [comment]
Hello! Do you know if this issue would be resolved in a near future? I'm loading some existing models (with parameter sharing like https://github.com/huggingface/transformers/blob/v4.17.0/src/transformers/models/bert/modeling_bert.py#L659) and might not be able to rewrite the models. My use case is computing per sample gradients or element-wise gradients with `jacrev` and `vmap`. Thank you!

**Comment by zou3519** [comment]
@xhan77 thanks for the interest! This issue is on our minds but we don't know if there's an easy fix we can apply. Thanks for the links to the huggingface models -- we agree those are important for us to support.

**Comment by zou3519** [comment]
@xhan77 this was a lot easier than I initially thought. We'll have a patch up later today or tomorrow that should fix the problem.

**Comment by zou3519** [comment]
@xhan77 https://github.com/pytorch/functorch/pull/620 fixed this. The fix will be in our next minor release (to come in < 1 month).

To use it earlier, here are some options:

(1) install the release branch from source (works with PyTorch 1.11)
```
# Makes the build go faster
pip install ninja
# Takes around 5-6 minutes to install
pip install --user "git+https://github.com/pytorch/functorch.git@release/0.1"
```
(2) Install functorch development branch from source (requires a PyTorch nightly binary)
https://github.com/pytorch/functorch#installing-functorch-from-source

Please let us know if your problem still persists and if there's anything else we can help out with!

**Comment by zou3519** [comment]
@fuzihaofzh we're planning on doing another release sometime next week, but let's see if we can solve your build-from-source problem
